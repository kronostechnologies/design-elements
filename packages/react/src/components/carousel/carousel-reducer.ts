export interface CarouselState {
    active: number;
    desired: number;
    offset: number;
    length: number;
    loop: boolean;
}

interface CarouselNextAction {
    type: 'next';
}

interface CarouselPrevAction {
    type: 'previous';
}

interface CarouselJumpAction {
    type: 'jump';
    desired: number;
}

interface CarouselDoneAction {
    type: 'done';
}

interface CarouselDragAction {
    type: 'drag';
    offset: number;
}

export type CarouselAction =
    | CarouselJumpAction
    | CarouselNextAction
    | CarouselPrevAction
    | CarouselDragAction
    | CarouselDoneAction;

export function carouselReducer(state: CarouselState, action: CarouselAction): CarouselState {
    switch (action.type) {
        case 'jump':
            return {
                ...state,
                desired: action.desired,
            };
        case 'next':
            return {
                ...state,
                desired: next(state.length, state.active),
            };
        case 'previous':
            return {
                ...state,
                desired: previous(state.length, state.active),
            };
        case 'done':
            return {
                ...state,
                offset: NaN,
                active: state.desired,
            };
        case 'drag':
            return {
                ...state,
                offset: action.offset,
            };
        default:
            return state;
    }
}

function previous(length: number, current: number): number {
    return (current - 1 + length) % length;
}

function next(length: number, current: number): number {
    return (current + 1) % length;
}
