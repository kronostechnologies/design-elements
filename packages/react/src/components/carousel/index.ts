export { Carousel } from './carousel';
export type { CarouselProps } from './carousel';
