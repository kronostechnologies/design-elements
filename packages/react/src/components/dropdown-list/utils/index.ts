export * from './dropdown-list-utils';
