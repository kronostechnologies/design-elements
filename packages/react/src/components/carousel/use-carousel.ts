// Source : https://gist.github.com/FlorianRappl/fee731eea985d983fc48d10c648ecb17

import { clamp } from '@design-elements/utils/math';
import { CSSProperties, Dispatch, useCallback, useEffect, useReducer } from 'react';
import { SwipeableHandlers, SwipeEventData, useSwipeable } from 'react-swipeable';
import { CarouselAction, carouselReducer, CarouselState } from './carousel-reducer';

type PreviousDirection = -1;
type NextDirection = 1;
type CarouselDirection = NextDirection | PreviousDirection;

interface CarouselHandlers {
    mouse: SwipeableHandlers;

    onNext(): void;

    onPrevious(): void;
}

export interface UseCarouselResponse {
    active: number;
    handlers: CarouselHandlers;
    style: CSSProperties;

    setActive(desired: number): void;
}

export interface UseCarouselOptions {
    autoTransitionDelay?: number;
    initial?: number;
    length: number;
    loop?: boolean;
    trackMouse?: boolean;
    trackTouch?: boolean;
    transitionTime?: number;
}

const FORWARD = 1;
const BACKWARD = -1;
const PADDING_SLIDES = 2;

function canSwipe(e: SwipeEventData, state: CarouselState): boolean {
    const direction = Math.sign(-e.deltaX);
    const isDraggingLastForward = state.active === state.length - 1 && direction === FORWARD;
    const isDraggingFirstBackward = state.active === 0 && direction === BACKWARD;

    return state.loop || (!state.loop && !isDraggingFirstBackward && !isDraggingLastForward);
}

function getClientWidth(target: EventTarget | null): number {
    return (target as HTMLElement).clientWidth;
}

function getSlideWidth(target: EventTarget | null, slidesCount: number): number {
    return getClientWidth(target) / (slidesCount + PADDING_SLIDES);
}

function threshold(target: EventTarget | null, slidesCount: number): number {
    if (target === null) {
        return 0;
    }

    return getSlideWidth(target, slidesCount) / 3;
}

function onSwiped(
    e: SwipeEventData,
    dispatch: Dispatch<CarouselAction>,
    state: CarouselState,
    direction: CarouselDirection,
): void {
    if (canSwipe(e, state)) {
        const delta = direction * -e.deltaX;
        if (delta >= threshold(e.event.currentTarget, state.length)) {
            dispatch({ type: direction === FORWARD ? 'next' : 'previous' });
        } else {
            dispatch({
                type: 'drag',
                offset: 0,
            });
        }
    }
}

function updateStyleForDraggingTransition(state: CarouselState, style: CSSProperties, transitionTime: number): void {
    if (state.offset !== 0) {
        // eslint-disable-next-line no-param-reassign
        style.transform = `translateX(${state.offset}px)`;
    } else {
        // eslint-disable-next-line no-param-reassign
        style.transition = `transform ${transitionTime}ms cubic-bezier(0.68, -0.55, 0.265, 1.55)`;
    }
}

interface NavigationTransition {
    distance: number;
    directionFromActive: CarouselDirection;
}

function getNavigationTransition(state: CarouselState): NavigationTransition {
    let distance: number;
    let directionFromActive: CarouselDirection;
    if (state.loop && state.desired === 0 && state.active > state.length / 2) {
        distance = state.length - state.active;
        directionFromActive = BACKWARD;
    } else if (state.loop && state.desired === state.length - 1 && state.active < Math.floor(state.length / 2)) {
        distance = state.length - (state.desired - state.active);
        directionFromActive = FORWARD;
    } else {
        distance = Math.abs(state.desired - state.active);
        directionFromActive = Math.sign(state.active - state.desired) as CarouselDirection;
    }

    return { distance, directionFromActive };
}

function updateStyleForTransitionToActive(state: CarouselState, style: CSSProperties, transitionTime: number): void {
    const directionFromDrag = Math.sign(state.offset || 0);
    const { distance, directionFromActive }: NavigationTransition = getNavigationTransition(state);

    let shift: number;
    if (directionFromDrag) {
        shift = 100 * directionFromDrag / (state.length + 2);
    } else {
        shift = 100 * directionFromActive / (state.length + 2) * distance;
    }

    // eslint-disable-next-line no-param-reassign
    style.transition = `transform ${transitionTime}ms ease`;
    // eslint-disable-next-line no-param-reassign
    style.transform = `translateX(${shift}%)`;
}

export function useCarousel(
    {
        transitionTime = 0,
        initial = 0,
        autoTransitionDelay,
        length,
        loop = false,
        trackMouse = false,
        trackTouch = true,
    }: UseCarouselOptions,
): UseCarouselResponse {
    const initialCarouselState: CarouselState = {
        active: initial,
        desired: initial,
        offset: 0,
        length,
        loop,
    };
    const [state, dispatch] = useReducer(carouselReducer, initialCarouselState);
    const swipeableHandlers: SwipeableHandlers = useSwipeable({
        onSwiping(e: SwipeEventData): void {
            if (canSwipe(e, state)) {
                const maxOffset = getSlideWidth(e.event.currentTarget, state.length);
                dispatch({
                    type: 'drag',
                    offset: clamp(-maxOffset, maxOffset, e.deltaX),
                });
            }
        },
        onSwipedLeft(e: SwipeEventData): void {
            onSwiped(e, dispatch, state, FORWARD);
        },
        onSwipedRight(e: SwipeEventData): void {
            onSwiped(e, dispatch, state, BACKWARD);
        },
        trackMouse,
        trackTouch,
    });

    const previousHandler: () => void = useCallback(() => {
        if (state.loop || (!state.loop && state.active > 0)) {
            dispatch({ type: 'previous' });
        }
    }, [state.loop, state.active]);

    const nextHandler: () => void = useCallback(() => {
        if (state.loop || (!state.loop && state.active < state.length - 1)) {
            dispatch({ type: 'next' });
        }
    }, [state.active, state.loop, state.length]);

    useEffect(() => {
        if (autoTransitionDelay !== undefined && Number.isFinite(autoTransitionDelay)) {
            const delay = Math.max(autoTransitionDelay, transitionTime);
            const id = setTimeout(nextHandler, delay);
            return () => clearTimeout(id);
        }
        return undefined;
    }, [transitionTime, nextHandler, autoTransitionDelay, state.offset, state.active]);

    useEffect(() => {
        const id = setTimeout(() => dispatch({ type: 'done' }), transitionTime);
        return () => clearTimeout(id);
    }, [state.desired, transitionTime]);

    const style: CSSProperties = {
        marginLeft: `-${(state.active + 1) * 100}%`,
        transform: 'translateX(0)',
        width: `${100 * (length + 2)}%`,
    };

    if (state.desired !== state.active) {
        updateStyleForTransitionToActive(state, style, transitionTime);
    } else if (!Number.isNaN(state.offset)) {
        updateStyleForDraggingTransition(state, style, transitionTime);
    }

    return {
        active: state.active,
        handlers: {
            mouse: swipeableHandlers,
            onPrevious: previousHandler,
            onNext: nextHandler,
        },
        setActive: (desired) => dispatch({ type: 'jump', desired }),
        style,
    };
}
