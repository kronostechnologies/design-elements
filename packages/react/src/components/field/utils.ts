import { ReactNode } from 'react';
import { devConsole } from '../../utils/dev-console';
import { joinStrings } from '../../utils/string';

interface FieldIds {
    label?: string;
    hint?: string;
    invalid?: string;
}

export function getSlotId(prop: ReactNode, fieldId: string, propName: string): string | undefined {
    return prop ? `${fieldId}_${propName}` : undefined;
}

export function getSlotIds(
    fieldId: string,
    label?: ReactNode,
    hint?: ReactNode,
    invalid?: ReactNode,
): FieldIds {
    return {
        label: getSlotId(label, fieldId, 'label'),
        hint: getSlotId(hint, fieldId, 'hint'),
        invalid: getSlotId(invalid, fieldId, 'invalid'),
    };
}

export const ARIA_LABEL_WARNING = ''
    + 'label, aria-label and aria-labelledby are mutually exclusive properties. Only one should be provided.';

export const validateAriaLabels = (
    label?: string,
    ariaLabel?: string,
    ariaLabelledby?: string,
): boolean => {
    const labelCount = [!!label, !!ariaLabel, !!ariaLabelledby].filter(Boolean).length;
    if (labelCount === 0 || labelCount > 1) {
        devConsole.warn(ARIA_LABEL_WARNING);
        return false;
    }
    return true;
};

export function getAriaLabel(
    label?: string,
    ariaLabel?: string,
    ariaLabelledby?: string,
): string | undefined {
    return validateAriaLabels(label, ariaLabel, ariaLabelledby) ? ariaLabel : undefined;
}

export function getAriaLabelledby(
    { label }: FieldIds,
    additionalLabelledby?: string,
): string | undefined {
    return joinStrings(label, additionalLabelledby);
}

export function getAriaDescribedby(
    { invalid, hint }: FieldIds,
    additionalDescribedby?: string,
): string | undefined {
    return joinStrings(invalid, hint, additionalDescribedby);
}
