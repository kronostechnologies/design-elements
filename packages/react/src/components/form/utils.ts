import { devConsole } from '../../utils/dev-console';

interface FormFieldIds {
    label?: string;
    hint?: string;
    invalid?: string;
}

function makeSlotId(prop: React.ReactNode, formFieldId: string, propName: string): string | undefined {
    if (!prop) {
        return undefined;
    }

    return `${formFieldId}_${propName}`;
}

export function getSlotIds(
    formFieldId: string,
    label?: React.ReactNode,
    hint?: React.ReactNode,
    invalid?: React.ReactNode,
): FormFieldIds {
    return {
        label: makeSlotId(label, formFieldId, 'label'),
        hint: makeSlotId(hint, formFieldId, 'hint'),
        invalid: makeSlotId(invalid, formFieldId, 'invalid'),
    };
}

export const ARIA_LABEL_WARNING = ''
    + 'label, aria-label and aria-labelledby are mutually exclusive properties. Only one should be provided.';

export const validateAriaLabels = (
    label?: string,
    ariaLabel?: string,
    ariaLabelledby?: string,
): boolean => {
    const labelCount = [!!label, !!ariaLabel, !!ariaLabelledby].filter(Boolean).length;
    if (labelCount === 0 || labelCount > 1) {
        devConsole.warn(ARIA_LABEL_WARNING);
        return false;
    }
    return true;
};

export function joinStrings(...strings: Array<string | undefined>): string | undefined {
    const filteredStrings = strings.filter(Boolean);
    return filteredStrings.length > 0 ? filteredStrings.join(' ') : undefined;
}

export function getAriaLabel(
    label?: string,
    ariaLabel?: string,
    ariaLabelledby?: string,
): string | undefined {
    return validateAriaLabels(label, ariaLabel, ariaLabelledby) ? ariaLabel : undefined;
}

export function getAriaLabelledBy(
    { label }: FormFieldIds,
    additionalLabelledBy?: string,
): string | undefined {
    return joinStrings(label, additionalLabelledBy);
}

export function getAriaDescribedBy(
    { invalid, hint }: FormFieldIds,
    additionalDescribedBy?: string,
): string | undefined {
    return joinStrings(invalid, hint, additionalDescribedBy);
}
