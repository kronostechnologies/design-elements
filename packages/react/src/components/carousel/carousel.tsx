import { AriaAttributes, FunctionComponent, PropsWithChildren, ReactElement, ReactNode } from 'react';
import styled, { DefaultTheme, StyledComponent } from 'styled-components';
import { useId } from '../../hooks/use-id';
import { focus } from '../../utils/css-state';
import { IconButton } from '../buttons';
import { useCarousel } from './use-carousel';

const Main = styled.section`
    align-items: center;
    display: flex;
    word-break: break-word;
`;

const Content = styled.div`
    height: 100%;
    margin: 0;
    overflow: hidden;
    padding: 0;
    width: 100%;
`;

interface SlideProps {
    index: number;
}

const Slide: StyledComponent<'div', DefaultTheme, SlideProps> = styled.div.attrs<SlideProps>({
    role: 'group',
    'aria-roledescription': 'slide',
})`
    flex-grow: 1;
    height: 100%;
    width: 100%;
`;

const Slides = styled.div.attrs({ 'aria-live': 'polite' })`
    align-items: center;
    display: flex;
`;

const Dots = styled.div`
    box-sizing: border-box;
    display: flex;
    justify-content: center;
    list-style: none;
    margin: 0;
    padding: 4px;
`;

interface DotProps {
    active?: boolean;
}

const Dot = styled.button<DotProps>`
    background-color: ${({ active, theme }) => (active ? theme.component['carousel-dot-selected-background-color'] : theme.component['carousel-dot-unselected-background-color'])};
    border: 1px solid  ${({ theme }) => theme.component['carousel-dot-border-color']};
    border-radius: 50%;
    box-sizing: border-box;
    display: inline-block;
    height: 12px;
    margin-bottom: 4px;
    width: 12px;

    ${focus};

    & + & {
        margin-left: var(--spacing-1x);
    }
`;

interface NavigationButtonProps {
    disabled?: boolean;
}

const NavigationButton = styled(IconButton)<NavigationButtonProps>`
    visibility: ${({ disabled }) => (disabled ? 'hidden' : 'visible')};

    ${focus};
`;

export interface CarouselProps extends Pick<AriaAttributes, 'aria-label'> {
    autoTransitionDelay?: number;
    /**
     * Anything React can render. Each root element will be rendered as a slide.
     */
    children: ReactNode;
    className?: string;
    header?: ReactNode;
    id?: string;
    initialSlide?: number;
    loop?: boolean;
    transitionTime?: number;
    withArrows?: boolean;
}

function getChildrenAsArray(children: ReactNode): ReactNode[] {
    if (!children) {
        return [];
    }
    if (Array.isArray(children)) {
        return children;
    }
    return [children];
}

export const Carousel: FunctionComponent<PropsWithChildren<CarouselProps>> = ({
    autoTransitionDelay,
    className,
    id,
    initialSlide = 0,
    children,
    header = null,
    loop = false,
    transitionTime = 400,
    withArrows = true,
    ...props
}) => {
    const carouselId = useId(id);

    const childrenAsArray = getChildrenAsArray(children);
    const slidesCount = childrenAsArray.length;
    const {
        active, setActive, handlers, style,
    } = useCarousel({
        autoTransitionDelay,
        initial: initialSlide,
        length: slidesCount,
        loop,
        transitionTime,
    });

    function renderDots(): ReactElement {
        return (
            <Dots data-testid="carousel-dots">
                {childrenAsArray.map((_, i) => {
                    const isActive = i === active;

                    return (
                        <Dot
                            data-testid={`carousel-dot-${i}`}
                            key={i /* eslint-disable-line react/no-array-index-key */}
                            active={isActive}
                            aria-controls={carouselId}
                            aria-disabled={isActive}
                            onClick={() => setActive(i)}
                        />
                    );
                })}
            </Dots>
        );
    }

    function renderSlides(): ReactElement {
        return (
            // eslint-disable-next-line react/jsx-props-no-spreading
            <Slides id={carouselId} data-testid="carousel-slides" {...handlers.mouse} style={style} aria-live="polite">
                <Slide
                    index={-1}
                    data-testid="carousel-slide--1"
                    aria-hidden
                >
                    {childrenAsArray[slidesCount - 1]}
                </Slide>

                {childrenAsArray.map((child, i) => (
                    <Slide
                        index={i}
                        data-testid={`carousel-slide-${i}`}
                        aria-hidden={i !== active}
                        key={i /* eslint-disable-line react/no-array-index-key */}
                        aria-label={`${i + 1} of ${slidesCount}`}
                    >
                        {child}
                    </Slide>
                ))}

                <Slide index={slidesCount} data-testid={`carousel-slide-${slidesCount}`} aria-hidden>
                    {childrenAsArray[0]}
                </Slide>
            </Slides>
        );
    }

    const disablePrevious = !loop && active === 0;
    const disableNext = !loop && active === slidesCount - 1;
    return (
        <Main className={className} aria-label={props['aria-label']} aria-roledescription="carousel">
            {withArrows && (
                <NavigationButton
                    buttonType="tertiary"
                    onClick={!disablePrevious ? handlers.onPrevious : undefined}
                    disabled={disablePrevious}
                    data-testid="carousel-previous"
                    aria-label="previous"
                    aria-controls={carouselId}
                    type="button"
                    iconName="arrowLeft"
                />
            )}

            <Content>
                {header}

                {renderSlides()}

                {renderDots()}
            </Content>

            {withArrows && (
                <NavigationButton
                    buttonType="tertiary"
                    onClick={!disableNext ? handlers.onNext : undefined}
                    disabled={disableNext}
                    data-testid="carousel-next"
                    aria-label="next"
                    aria-controls={carouselId}
                    type="button"
                    iconName="arrowRight"
                />
            )}
        </Main>
    );
};
