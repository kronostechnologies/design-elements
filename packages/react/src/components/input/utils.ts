import { InputFieldProps, InputProps } from './types';
import { FieldContainerProps } from '../field/types';

interface ParsedInputFieldProps {
    inputProps: InputProps;
    fieldControlledProps: FieldContainerProps;
}

export function parseInputFieldProps(
    props: InputFieldProps,
): ParsedInputFieldProps {
    const {
        ariaLabel,
        ariaLabelledby,
        ariaDescribedby,
        valid,
        id,
        disabled,
        required,
        className,
        noMargin,
        label,
        hint,
        validationErrorMessage,
        noInvalidFieldIcon,
        tooltip,
        ...inputProps
    } = props;

    const fieldControlledProps: FieldContainerProps = {
        id,
        ariaLabel,
        ariaLabelledby,
        ariaDescribedby,
        valid,
        className,
        noMargin,
        label,
        hint,
        validationErrorMessage,
        noInvalidFieldIcon,
        tooltip,
        disabled,
        required,
    };

    return { inputProps, fieldControlledProps };
}
