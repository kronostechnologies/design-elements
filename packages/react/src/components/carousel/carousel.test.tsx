import { findByTestId, getByTestId } from '@design-elements/test-utils/enzyme-selectors';
import { shallow } from 'enzyme';
import React, { ReactNode } from 'react';
import { mocked } from 'ts-jest/utils';
import { renderWithTheme } from '../../test-utils/theme-wrapped';
import { Carousel } from './carousel';
import { useCarousel, UseCarouselResponse } from './use-carousel';

jest.mock('./use-carousel');

describe('Carousel', () => {
    let slides: ReactNode[];
    let numberOfSlides: number;
    let useCarouselResponse: UseCarouselResponse;

    beforeEach(() => {
        slides = [
            <div key={1}>Slide 1</div>,
            <div key={2}>Slide 2</div>,
            <div key={3}>Slide 3</div>,
            <div key={4}>Slide 4</div>,
            <div key={5}>Slide 5</div>,
        ];
        numberOfSlides = slides.length;

        useCarouselResponse = {
            active: 0,
            handlers: {
                mouse: { ref: jest.fn(), onMouseDown: jest.fn() },
                onNext: jest.fn(),
                onPrevious: jest.fn(),
            },
            setActive: jest.fn(),
            style: {},
        };
        mocked(useCarousel).mockReturnValue(useCarouselResponse);
    });

    it('should match snapshot', () => {
        const wrapper = renderWithTheme(<Carousel>{slides}</Carousel>);

        expect(wrapper).toMatchSnapshot();
    });

    describe('Padding slides', () => {
        it('should add padding slides so looping does not jump', () => {
            const wrapper = shallow(<Carousel>{slides}</Carousel>);

            const renderedSlides = getByTestId(wrapper, 'carousel-slides', '$');

            expect(renderedSlides.children().length).toBe(numberOfSlides + 2);
        });

        it('should add last slide as padding at the beginning', () => {
            const wrapper = shallow(<Carousel>{slides}</Carousel>);

            const paddingSlide = getByTestId(wrapper, 'carousel-slide--1');

            expect(paddingSlide.text()).toBe('Slide 5');
        });

        it('should add first slide as padding at the end', () => {
            const wrapper = shallow(<Carousel>{slides}</Carousel>);

            const paddingSlide = getByTestId(wrapper, 'carousel-slide-5');

            expect(paddingSlide.text()).toBe('Slide 1');
        });
    });

    describe('Navigation', () => {
        it('should go forward when next is clicked', () => {
            const wrapper = shallow(<Carousel>{slides}</Carousel>);

            getByTestId(wrapper, 'carousel-next').simulate('click');

            expect(useCarouselResponse.handlers.onNext).toHaveBeenCalledTimes(1);
        });

        it('should not go forward when next is clicked and last slide is active', () => {
            useCarouselResponse.active = numberOfSlides - 1;
            const wrapper = shallow(<Carousel>{slides}</Carousel>);

            const nextWrapper = getByTestId(wrapper, 'carousel-next').simulate('click');

            expect(nextWrapper.props().disabled).toBe(true);
            expect(useCarouselResponse.handlers.onNext).not.toHaveBeenCalled();
        });

        it('should go backward when previous is clicked', () => {
            useCarouselResponse.active = 3;
            const wrapper = shallow(<Carousel>{slides}</Carousel>);

            getByTestId(wrapper, 'carousel-previous').simulate('click');

            expect(useCarouselResponse.handlers.onPrevious).toHaveBeenCalledTimes(1);
        });

        it('should not go backward when previous is clicked and first slide is active', () => {
            useCarouselResponse.active = 0;
            const wrapper = shallow(<Carousel>{slides}</Carousel>);

            const previousWrapper = getByTestId(wrapper, 'carousel-previous').simulate('click');

            expect(previousWrapper.props().disabled).toBe(true);
            expect(useCarouselResponse.handlers.onPrevious).not.toHaveBeenCalled();
        });

        describe('Loop', () => {
            it('should go backward when previous is clicked and first slide is active', () => {
                useCarouselResponse.active = 0;
                const wrapper = shallow(<Carousel loop={true}>{slides}</Carousel>);

                getByTestId(wrapper, 'carousel-previous').simulate('click');

                expect(useCarouselResponse.handlers.onPrevious).toHaveBeenCalled();
            });

            it('should go forward when next is clicked and last slide is active', () => {
                useCarouselResponse.active = numberOfSlides - 1;
                const wrapper = shallow(<Carousel loop={true}>{slides}</Carousel>);

                getByTestId(wrapper, 'carousel-next').simulate('click');

                expect(useCarouselResponse.handlers.onNext).toHaveBeenCalled();
            });
        });
    });

    describe('Navigation with Dots', () => {
        it('should display one dot per slide', () => {
            const wrapper = shallow(<Carousel>{slides}</Carousel>);

            const dots = findByTestId(wrapper, 'carousel-dot-', '^');

            expect(dots.length).toBe(5);
        });

        it('should go to slide when dot is clicked', () => {
            const wrapper = shallow(<Carousel>{slides}</Carousel>);

            getByTestId(wrapper, 'carousel-dot-2').simulate('click');

            expect(useCarouselResponse.setActive).toHaveBeenCalledWith(2);
        });

        it('should set dot as active when slide is active', () => {
            useCarouselResponse.active = 3;
            const wrapper = shallow(<Carousel>{slides}</Carousel>);

            const dot = getByTestId(wrapper, 'carousel-dot-3');

            expect(dot.props().active).toBe(true);
        });
    });
});
