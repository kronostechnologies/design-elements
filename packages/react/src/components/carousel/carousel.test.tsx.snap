// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel should match snapshot 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  word-break: break-word;
}

.c2 {
  height: 100%;
  margin: 0;
  overflow: hidden;
  padding: 0;
  width: 100%;
}

.c4 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  height: 100%;
  width: 100%;
}

.c3 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c5 {
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  list-style: none;
  margin: 0;
  padding: 0;
}

.c7 {
  background-color: #006296;
  border: 1px solid transparent;
  border-radius: 50%;
  box-sizing: border-box;
  display: inline-block;
  height: 12px;
  width: 12px;
  margin-bottom: 2px;
}

.c7:not(:focus) {
  -webkit-transition: all .25s ease-in-out;
  transition: all .25s ease-in-out;
  box-shadow: none;
  outline: 2px solid transparent;
  outline-offset: -2px;
}

.c7:focus {
  -webkit-transition: all .25s ease-in-out;
  transition: all .25s ease-in-out;
  box-shadow: 0 0 0 2px #006296;
  outline: 2px solid #84C6EA;
  outline-offset: -2px;
}

.c6 + .c6 {
  margin-left: var(--spacing-1x);
}

.c8 {
  background-color: #DBDEE1;
  border: 1px solid transparent;
  border-radius: 50%;
  box-sizing: border-box;
  display: inline-block;
  height: 12px;
  width: 12px;
  margin-bottom: 2px;
}

.c8:not(:focus) {
  -webkit-transition: all .25s ease-in-out;
  transition: all .25s ease-in-out;
  box-shadow: none;
  outline: 2px solid transparent;
  outline-offset: -2px;
}

.c8:focus {
  -webkit-transition: all .25s ease-in-out;
  transition: all .25s ease-in-out;
  box-shadow: 0 0 0 2px #006296;
  outline: 2px solid #84C6EA;
  outline-offset: -2px;
}

.c6 + .c6 {
  margin-left: var(--spacing-1x);
}

.c1 {
  border: 1px solid transparent;
  height: 34px;
  visibility: hidden;
  width: 34px;
}

.c1:not(:focus) {
  -webkit-transition: all .25s ease-in-out;
  transition: all .25s ease-in-out;
  box-shadow: none;
  outline: 2px solid transparent;
  outline-offset: -2px;
}

.c1:focus {
  -webkit-transition: all .25s ease-in-out;
  transition: all .25s ease-in-out;
  box-shadow: 0 0 0 2px #006296;
  outline: 2px solid #84C6EA;
  outline-offset: -2px;
}

.c9 {
  border: 1px solid transparent;
  height: 34px;
  visibility: visible;
  width: 34px;
}

.c9:not(:focus) {
  -webkit-transition: all .25s ease-in-out;
  transition: all .25s ease-in-out;
  box-shadow: none;
  outline: 2px solid transparent;
  outline-offset: -2px;
}

.c9:focus {
  -webkit-transition: all .25s ease-in-out;
  transition: all .25s ease-in-out;
  box-shadow: 0 0 0 2px #006296;
  outline: 2px solid #84C6EA;
  outline-offset: -2px;
}

<section
  aria-roledescription="carousel"
  class="c0"
>
  <button
    aria-controls="uuid1"
    aria-label="previous"
    class="c1"
    data-testid="carousel-previous"
    disabled=""
    type="button"
  >
    <svg
      color="#60666E"
      focusable="false"
      height="100%"
      width="100%"
    />
  </button>
  <div
    class="c2"
  >
    <div
      aria-live="polite"
      class="c3"
      data-testid="carousel-slides"
      id="uuid1"
    >
      <div
        aria-hidden="true"
        aria-roledescription="slide"
        class="c4"
        data-testid="carousel-slide--1"
        role="group"
      >
        <div>
          Slide 5
        </div>
      </div>
      <div
        aria-hidden="false"
        aria-label="1 of 5"
        aria-roledescription="slide"
        class="c4"
        data-testid="carousel-slide-0"
        role="group"
      >
        <div>
          Slide 1
        </div>
      </div>
      <div
        aria-hidden="true"
        aria-label="2 of 5"
        aria-roledescription="slide"
        class="c4"
        data-testid="carousel-slide-1"
        role="group"
      >
        <div>
          Slide 2
        </div>
      </div>
      <div
        aria-hidden="true"
        aria-label="3 of 5"
        aria-roledescription="slide"
        class="c4"
        data-testid="carousel-slide-2"
        role="group"
      >
        <div>
          Slide 3
        </div>
      </div>
      <div
        aria-hidden="true"
        aria-label="4 of 5"
        aria-roledescription="slide"
        class="c4"
        data-testid="carousel-slide-3"
        role="group"
      >
        <div>
          Slide 4
        </div>
      </div>
      <div
        aria-hidden="true"
        aria-label="5 of 5"
        aria-roledescription="slide"
        class="c4"
        data-testid="carousel-slide-4"
        role="group"
      >
        <div>
          Slide 5
        </div>
      </div>
      <div
        aria-hidden="true"
        aria-roledescription="slide"
        class="c4"
        data-testid="carousel-slide-5"
        role="group"
      >
        <div>
          Slide 1
        </div>
      </div>
    </div>
    <div
      class="c5"
      data-testid="carousel-dots"
    >
      <button
        aria-controls="uuid1"
        aria-disabled="true"
        class="c6 c7"
        data-testid="carousel-dot-0"
      />
      <button
        aria-controls="uuid1"
        aria-disabled="false"
        class="c6 c8"
        data-testid="carousel-dot-1"
      />
      <button
        aria-controls="uuid1"
        aria-disabled="false"
        class="c6 c8"
        data-testid="carousel-dot-2"
      />
      <button
        aria-controls="uuid1"
        aria-disabled="false"
        class="c6 c8"
        data-testid="carousel-dot-3"
      />
      <button
        aria-controls="uuid1"
        aria-disabled="false"
        class="c6 c8"
        data-testid="carousel-dot-4"
      />
    </div>
  </div>
  <button
    aria-controls="uuid1"
    aria-label="next"
    class="c9"
    data-testid="carousel-next"
    type="button"
  >
    <svg
      color="#60666E"
      focusable="false"
      height="100%"
      width="100%"
    />
  </button>
</section>
`;
